Initial enterprise linked domains

orangejenny opened the pull request:
##### SUMMARY
https://trello.com/c/wRKqEipZ/23-enterprise-web-users-want-to-automatically-have-the-same-level-of-access-across-all-counties

##### FEATURE FLAG
New flag: Give this domain enterprise functionality as related to any domains linked to it.

##### PRODUCT DESCRIPTION
This adds a feature flag to turn on for a linked domain's **master** domain. With the flag on, any web users who are domain admins in the master domain will automatically have admin privileges on all domains linked to that master.

Any web users who are members of the master domain but not admins will be automatically granted access to linked domains, with the same level of access they have on the master.

This is handled similar to superuser access: these users will be able to log into and modify the linked domains, but they are not considered members of the linked domain, so they will not show up in places like the linked domains' lists of web users.

These users will see an "Enterprise Projects" link in the domain dropdown:
<img width="301" alt="Screen Shot 2020-05-11 at 7 23 58 PM" src="https://user-images.githubusercontent.com/1486591/81621674-0201dd00-93bd-11ea-965d-1c1f611808e6.png">

This link goes to the pre-existing domain selection page, where you get sent if you log into HQ and we can't automatically redirect you to your most recently used domain. The select page has a new section for "Enterprise Projects":
<img width="1108" alt="Screen Shot 2020-05-11 at 7 25 24 PM" src="https://user-images.githubusercontent.com/1486591/81621742-307fb800-93bd-11ea-9209-52cf9cbda06c.png">

This new flag does not work with remote linked domains. It also does not work with location restrictions.

This has nothing to do with the current "enterprise admins" that are configured via billing. It is built on top of linked domains.

fyi @dimagi/product 

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1294)

orangejenny committed:
Added feature flag

orangejenny committed:
Added linked domain admin-ship for admins of master domain

orangejenny committed:
Moved invitations above projects on domain selection page

Prevent this section from getting lost, as the new section may contain dozens of domains.

orangejenny committed:
Added linked project spaces to domain select page

orangejenny committed:
DRYed up code between domain select view and domain dropdown

orangejenny committed:
Added 'Enterprise Projects' link to dropdown

orangejenny committed:
Removed 'Have a new idea?' box

This is old and looks out of place. Checked with Dev it's okay to remove.

orangejenny committed:
Removed dead code

orangejenny committed:
Added test

orangejenny committed:
Added linked domains to dropdown/select even if user isn't an admin

orangejenny committed:
Updated linked domain permissions to mirror master domain permissions

orangejenny committed:
Added tests

calellowitz commented on corehq/apps/domain/templates/domain/select.html:
```diff
@@ -26,17 +20,19 @@ <h2>{% trans "Project Invitations" %}</h2>
-        {% if not hide_create_new_project %}
-          <aside class="well">
-            <h3>{% trans "Have a new idea?" %}</h3>
-            <a class="btn btn-primary" href="{% url "registration_domain" %}">{% trans "Create a Blank Project" %}</a>
```
does this mean you are no longer able to register a domain from that page?

orangejenny replied:
yes, you can now only register a domain from the domain dropdown

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -66,6 +73,36 @@ def select(request, do_not_redirect=False, next_view=None):
+            links = get_linked_domains(domain)
+            links = [link for link in links if link.linked_domain not in domain_names]
+            linked_domains = [Domain.get_by_name(link.linked_domain) for link in links]
+            linked_domain_links.extend([Link(
```
i dont have a better word to suggest, but the use of link in the sense of "domain linking" and the sense of "html link" is mildly confusing here, especially with names like `linked_domain_links` that encompass both

orangejenny replied:
Agreed this is confusing. Added fe0cc2197b1 to favor "enterprise" over "linked" when describing domains. Enterprise is also overloaded but then at least the same code isn't using "linked" and "link" in very different ways. Open to further feedback.

orangejenny commented:
Updated so that non-admin members of the master domain also get access to the linked domains, with the same permission set they have on the master. Updated PR description.

Ready for review.

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -66,6 +73,36 @@ def select(request, do_not_redirect=False, next_view=None):
+    linked_domain_links = []
+    domain_names = {d.name for d in domain_links}
+    for domain_obj in domain_links:
+        domain = domain_obj.name
```
This code switches between the word `domain` meaning the name and the object (and likewise sometimes using `domain_name` and sometimes using `domain_obj`, such that it is hard to keep track of what type the unadorned word is.

orangejenny replied:
I meant to use `domain` for all names and `domain_obj` for all objects. The only inconsistency I see is that `domain_names` could be called `domains` - fixing this in 5eeb7946a9f

calellowitz replied:
I think its slightly more confusing than just that one.
Line 81 uses `domains` for a list of objects. and `linked_domains` is also a list of objects in line 94. but the `linked_domain` attribute of `link` is a name, I think.
To add to that confusion, `domain_obj` here isn't actually a domain object, its a `Link` object, unlike the `domain_obj` in line 85, which I believe is an object of class `Domain`.
Anyway, I am not trying to be neurotic about these things. I was just definitely confused going through this code and had to keep referring back to where variables were defined to figure out what type they were.

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -688,14 +697,24 @@ def has_permission(self, domain, permission, data=None, restrict_global_admin=Fa
+        master_link = get_domain_master_link(domain)
+        if master_link and not master_link.is_remote:
+            if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(master_link.master_domain):
+                if _domain_membership_has_permission(self.get_domain_membership(master_link.master_domain)):
```
This means that a user will have all permissions they have in the master domain in all the linked ones, regardless of role in the master (ie not just admin)?

Is it a true statement to say that you can add permissions in a linked domain (by giving them an explicit role with new permissions), but not remove them, since it will fall back to checking the master if they do not have the permission in the linked domain?

That might be slightly confusing, if a user is given an intentionally restricted role in a given linked domain, but inherits a bunch of permissions from the master domain. It is at least worth documenting that somewhere, if that is true. 

If it solves the main use case, it might also be worth returning the value from the local domain membership if it exists, rather than just checking if the membership grants a permission, and moving on if not.

orangejenny replied:
What you're describing is how it works, yes. Let's offline with Dev about if that's how it should work.

orangejenny replied:
Updated in c5a96b38351

calellowitz reviewed (commented):
A few questions about the behavior (and some minor naming things). I left this as a comment not approval because I am curious about the answer to the last question, but it should not be considered blocking.

orangejenny committed:
Updated naming

orangejenny committed:
Renamed domain_names

orangejenny committed:
Updated to make linked domain role, if present, supersede master domain role

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -84,23 +84,23 @@ def get_domain_dropdown_links(couch_user, view_name="domain_homepage"):
         if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(domain):
             links = get_linked_domains(domain)
             links = [link for link in links if link.linked_domain not in domain_names]
             linked_domains = [Domain.get_by_name(link.linked_domain) for link in links]
```
nit: this could be `enterprise_domains` for consistency with the below

calellowitz commented on corehq/apps/domain/views/base.py:
```diff
@@ -84,23 +84,23 @@ def get_domain_dropdown_links(couch_user, view_name="domain_homepage"):
     for domain_obj in domain_links:
         domain = domain_obj.name
         if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(domain):
             links = get_linked_domains(domain)
```
nit: these are domain objects i think

calellowitz reviewed (approved):
Left a few more naming comments because I can't help myself. But this looks good (and better)

orangejenny committed:
Cleaned up naming

orangejenny commented:
@calellowitz Did some more naming cleanup in 847e3c20433

orangejenny committed:
Reduced code in MyProjectsList

orangejenny committed:
Removed CouchUser.projects

Grepping for "\.projects\b" turned up nothing.

orangejenny committed:
Updated login_and_domain_required

orangejenny committed:
Extracted _AuthorizableMixin.get_enterprise_membership

orangejenny committed:
Removed dead membership classes

orangejenny committed:
Removed 'old_couch_user' functionality

This is from 2011 and no longer referenced, see e0073d1d5199874a29dfbebd80e7b886ee21a961

orangejenny committed:
Added include_enterprise to is_member_of

orangejenny committed:
Added enterprise domains to checking for requring secure session

orangejenny commented:
@calellowitz Added more commits after testing and finding some more permissions-related functionality. Rebased a few times (sorry, forgot this was already being reviewed) but I didn't change any of the commit's you've already looked at (`Cleaned up naming` and earlier).

orangejenny committed:
Fixed None-related errors

orangejenny committed:
Fixed return value in _get_viewable_report_slugs

orangejenny committed:
Fixed get_domain_dropdown_links missing domains

Previous code was incorrectly conflating link names (which use domain display names)
with display names. Updated to gather all domains first, then map to links afterwards.

orangejenny committed:
Added 'Enterprise User' label for gear dropdown

orangejenny committed:
Fixed any possibility of 'Enterprise Projects' link redirecting

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -1610,8 +1596,8 @@ def _get_viewable_report_slugs(self, domain):
-            dm = self.get_domain_membership(domain)
-            return dm.viewable_reports() if dm else []
+            membership = self.get_domain_membership(domain) or self.get_enterprise_membership(domain)
+            return membership.viewable_reports() or []
```
can `membership` be `None`. it looks like the old code was guarding against that, and this no longer does

orangejenny replied:
Not following. If `membership` is `None`, this returns an empty list, in both the old code (1597) and the new code (1562). Am I missing something?

calellowitz replied:
if `membership` is `None`, wont `membership.viewable_reports()` throw an `AttributeError`

orangejenny replied:
offlined, resolved

calellowitz commented on corehq/apps/domain/decorators.py:
```diff
@@ -100,21 +101,28 @@ def call_view(): return view_func(req, domain_name, *args, **kwargs)
             return call_view()
-        elif couch_user.is_web_user() and domain_obj.allow_domain_requests:
+
+        from corehq.apps.linked_domain.dbaccessors import get_domain_master_link
+        master_link = get_domain_master_link(domain_obj.name)
+        if master_link and not master_link.is_remote:
+            if ENTERPRISE_LINKED_DOMAINS.enabled(master_link.master_domain):
+                return call_view()
```
I believe this is missing a check for membership/permission

orangejenny replied:
updated in cdd8e0fc93c

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -526,13 +526,24 @@ class Meta(object):
+            return True
+
+        if include_enterprise:
+            for domain in domains:
```
why loop over all domains, rather than call for the domain passed in?

orangejenny replied:
5131064418f

calellowitz commented on corehq/apps/users/models.py:
```diff
@@ -526,13 +526,24 @@ class Meta(object):
+                master_link = get_domain_master_link(domain)
+                if master_link and not master_link.is_remote:
+                    if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(master_link.master_domain):
+                        return self.get_domain_membership(master_link.master_domain)
```
Is it possible to get here multiple domains would end up here, maybe if you were part of multiple enterprises? It seems like the value returned depends on the order of domains, since it early exits as soon as it gets here.

orangejenny replied:
Shouldn't have been looping in the first place. Fixed.

calellowitz commented on corehq/middleware.py:
```diff
@@ -123,8 +124,19 @@ def _session_expired(timeout, activity, time):
+
+        from corehq.apps.linked_domain.dbaccessors import get_linked_domains
+        for domain in domains:
+            if toggles.ENTERPRISE_LINKED_DOMAINS.enabled(domain):
```
This means `secure_session` set on the enterprise applies to all linked domains?

orangejenny replied:
My thought was that the intention of this code is "return true if any domains the user has access to use secure sessions" and that enterprise users do have access to all linked domains. And that it's better to be more aggressive, rather than less aggressive, when it comes to secure sessions.

orangejenny replied:
I think [this comment](https://github.com/dimagi/commcare-hq/pull/10901#issuecomment-200444006) applies here.

calellowitz reviewed (changes requested):
Few small questions, some from our slack chat, but looks good overall

orangejenny committed:
Fixed 500 for mobile users with ICDS flag on

orangejenny committed:
Fixed unnecessary loop in _is_member_of

orangejenny committed:
Fixed login_and_domain_required check

orangejenny commented:
Requirements changed. Closing this.
